ecs-agent/wsclient/client_test.go

```diff
@@ -0,0 +1,400 @@
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package wsclient
+
+import (
+	"context"
+	"errors"
+	"io"
+	"net"
+	"net/url"
+	"os"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/aws/amazon-ecs-agent/ecs-agent/acs/model/ecsacs"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/wsclient/mock/utils"
+	mock_wsconn "github.com/aws/amazon-ecs-agent/ecs-agent/wsclient/wsconn/mock"
+	"github.com/golang/mock/gomock"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+
+	"github.com/gorilla/websocket"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+const dockerEndpoint = "/var/run/docker.sock"
+
+// Close closes the underlying connection. Implement Close() in this file
+// as ClientServerImpl doesn't implement it. This is needed by the
+// TestSetReadDeadline* tests
+func (cs *ClientServerImpl) Close() error {
+	return cs.Disconnect()
+}
+
+func TestClientProxy(t *testing.T) {
+	proxy_url := "127.0.0.1:1234"
+	os.Setenv("HTTP_PROXY", proxy_url)
+	defer os.Unsetenv("HTTP_PROXY")
+
+	types := []interface{}{ecsacs.AckRequest{}}
+	cs := getClientServer("http://www.amazon.com", types)
+	err := cs.Connect()
+	assert.Error(t, err)
+	assert.True(t, strings.Contains(err.Error(), proxy_url), "proxy not found: %s", err.Error())
+}
+
+// TestConcurrentWritesDontPanic will force a panic in the websocket library if
+// the implemented methods don't handle concurrency correctly
+// See https://godoc.org/github.com/gorilla/websocket#hdr-Concurrency
+func TestConcurrentWritesDontPanic(t *testing.T) {
+	closeWS := make(chan []byte)
+	defer close(closeWS)
+
+	mockServer, _, requests, _, _ := utils.GetMockServer(closeWS)
+	mockServer.StartTLS()
+	defer mockServer.Close()
+
+	var waitForRequests sync.WaitGroup
+	waitForRequests.Add(1)
+
+	go func() {
+		for i := 0; i < 20; i++ {
+			<-requests
+		}
+		waitForRequests.Done()
+	}()
+	req := ecsacs.AckRequest{Cluster: aws.String("test"), ContainerInstance: aws.String("test"), MessageId: aws.String("test")}
+
+	types := []interface{}{ecsacs.AckRequest{}}
+	cs := getClientServer(mockServer.URL, types)
+	require.NoError(t, cs.Connect())
+
+	executeTenRequests := func() {
+		for i := 0; i < 10; i++ {
+			assert.NoError(t, cs.MakeRequest(&req))
+		}
+	}
+
+	// Make requests from two separate routines to try and force a
+	// concurrent write
+	go executeTenRequests()
+	go executeTenRequests()
+
+	t.Log("Waiting for all 20 requests to succeed")
+	waitForRequests.Wait()
+}
+
+func getClientServer(url string, msgType []interface{}) *ClientServerImpl {
```

Comment:
Can we change this name to be `dummyClientServer` or `testClientServer` or similar? That seems a lot more clear to me.

Reply:
Sure. Done.